abi.html

```diff
@@ -4107,73 +4106,156 @@ <h5><a href="#dso-dtor-motivation"> 3.3.6.1 Motivation </a></h5>
+
+<p>The third argument, <code>d</code>, must be the value
+<code>&__dso_handle</code> for the DSO which defines the object;
+see below.  The first and second arguments may be chosen at the
+implementation's convenience.
 
-<ol type=A>
 <p>
-<li> Object construction:
+The runtime library will call <code>f(p)</code> when the unloading
+of DSO <code>d</code> is required, including when the process exits.
+The call will occur before the DSO is actually unloaded from the
+process.  Calls to global termination functions will be made in the
+reverse order of their registration.
+
 <p>
-After constructing a global (or local static) object,
-that will require destruction on exit,
-a termination function is <i>registered</i> as follows:
-<center><code>
-extern "C" int __cxa_atexit ( void (*f)(void *), void *p, void *d );
-</code></center>
-This registration, e.g. <code>__cxa_atexit(f,p,d)</code>,
-is intended to cause the call <code>f(p)</code> when DSO <code>d</code> is unloaded,
-before all such termination calls registered before this one.
-It returns zero if registration is successful, nonzero on failure.
+Note that global termination functions may be registered
+concurrently.  If the registrations of two separate termination
+functions are not well-ordered by the <i>strongly happens before</i>
+relation, the order in which they are called is unspecified.
 
 <p>
-The registration function is not called from within the constructor.
+<code>__cxa_atexit</code> returns zero if registration is successful,
+nonzero on failure.
+
+<a name="thread-local-dtor">
+<h5><a href="#thread-local-dtor">3.3.6.4 Destructors for thread-local objects</a></h5>
 
 <p>
-<li> User <code>atexit</code> calls:
+After constructing an object with thread storage duration
+that will require destruction on process or thread exit,
+a thread-local termination function is registered for the
+current thread by calling the following function:
+
+<code><pre>extern "C" int __cxa_thread_atexit(void (*f)(void *), void *p, void *d);</pre></code>
+
+<p>The third argument, <code>d</code>, must be the value
+<code>&__dso_handle</code> for the DSO which defines the object;
+see below.  The first and second arguments may be chosen at the
+implementation's convenience.
+
 <p>
-When the user registers exit functions with <code>atexit</code>,
-they should be registered with NULL parameters and DSO handles, i.e.
-<center><code>
-    __cxa_atexit ( f, NULL, NULL );
-</code></center>
-It is expected that implementations supporting both C and C++ will
-integrate this capability into the libc <code>atexit</code>
-implementation so that C-only DSOs will nevertheless interact with C++
-programs in a C++-standard-conforming manner.
-No user interface to <code>__cxa_atexit</code> is supported,
-so the user is not able to register an <code>atexit</code> function
-with a parameter or a home DSO.
+The runtime library will call <code>f(p)</code> if the
```

Comment:
This requires memory allocation, but there is no way to report memory allocation failure. Thread creation is probably a way more common occurrence than global initialization (despite `dlopen`), so fixing this for `__cxa_thread_atexit` seems more important than for `__cxa_atexit`.

One way to fix this is to have an out-of-spec mechanism by which the link editor communicates the maximum number of objects needing such registration to the run-time library. (With ELF, it would probably not involve symbols, but some other mechanis,.)

Reply:
I agree that the ABI would be better if the allocation for this wasn't dynamic.  A simpler solution than adding a new kind of runtime data collection would be for the compiler to allocate some fixed amount of additional global/thread-local memory for global/thread-local objects that need destruction, the same way that it allocates additional memory for a guard variable.  This memory would be wasted if it was eagerly allocated and the object was never touched, but so would the memory for the object itself; that seems like an acceptable cost to me.

However, `__cxa_thread_atexit` is not actually new ABI; it's been implemented for years, and the proposal just never made it into the document, for various not-very-good reasons.  So any new proposal here would have to be a v2 recommendation rather than an actual change to `__cxa_thread_atexit`.

`__cxa_thread_atexit` does in fact have a return value that it can use to report failure.  Neither GCC nor Clang seems to actually check it; I don't know if any compilers do.  Frankly, I'm not sure we're really *allowed* to fail once the initializer has successfully terminated; it would have to be by throwing an exception, and there's no provision for doing so.  That's probably the strongest argument for changing the ABI.